chain/client/src/sync/epoch.rs

```diff
@@ -74,21 +92,60 @@ impl EpochSync {
         //
         // The basic requirement for picking the target epoch is that its first block must be
         // final. That's just so that we don't have to deal with any forks. Therefore, it is
-        // sufficient to pick whatever epoch the current final block is in. However, because
-        // state sync also requires some previous headers to be available (depending on how
-        // many chunks were missing), it is more convenient to just pick the prior epoch, so
-        // that by the time the new node does state sync, it would have a whole epoch of headers
-        // available via header sync.
-        //
-        // In other words, we pick the target epoch to be the previous epoch of the final tip.
-        let tip = store
-            .get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?
-            .ok_or_else(|| Error::Other("Could not find tip".to_string()))?;
-        let next_next_epoch_id = tip.next_epoch_id;
-        // Last block hash of the target epoch is the same as the next next EpochId.
-        // That's a general property for Near's epochs.
+        // sufficient to pick whatever epoch the current final block is in. However, there are
+        // additional considerations:
+        //  - Because state sync also requires some previous headers to be available (depending
+        //    on how many chunks were missing), if we pick the most recent epoch, after the node
+        //    finishes epoch sync and transitions to state sync, it would not have these headers
+        //    before the epoch. Therefore, for this purpose, it's convenient for epoch sync to not
+        //    pick the most recent epoch. This would ensure that state sync has a whole epoch of
+        //    headers before the epoch it's syncing to.
+        //  - We also need to have enough block headers to check for transaction_validity_period.
+        //    Therefore, we need find the latest epoch for which we would have at least that many
+        //    headers.
+        let target_epoch_last_block_hash = {
```

Comment:
nit: making it a separate method would make it more readable to me.

Reply:
Done